File: packages/claude-code-toolbox/src/agents/runtime/protocols/roundRobin.ts
/**
 * Round-robin protocol — custom runtime.
 *
 * Agents speak in order, each seeing the previous agents' latest messages.
 * Wraps around until `maxTurns` total turns have been consumed or the signal
 * is aborted. Each agent emits a `message {from, to: next}` at the end of its
 * turn so the transcript shows the handoff clearly.
 */
import type { Protocol } from "../runtimeTypes";
import { emitMessage, findAgentByName, makeTotals, runOneTurn } from "./shared";

export const roundRobin: Protocol = async (ctx) => {
  const totals = makeTotals();
  const roster = ctx.team.agents
    .map((n) => findAgentByName(ctx.agents, n))
    .filter((a): a is NonNullable<typeof a> => a !== undefined);
  if (!roster.length) {
    return { status: "error", totals };
  }
  let history: { agent: string; text: string }[] = [];
  const max = Math.min(ctx.team.maxTurns || 10, roster.length * 8);
  for (let turn = 1; turn <= max && !ctx.signal.aborted; turn++) {
    const agent = roster[(turn - 1) % roster.length];
    const transcript = history
      .slice(-6)
      .map((h, i) => `${i + 1}. ${h.agent}: ${h.text}`)
      .join("\n");
    const prompt = [
      `Team round-robin turn ${turn}/${max}. Team task:`,
      "",
      ctx.userPrompt,
      "",
      transcript ? `Recent transcript:\n${transcript}` : "(you speak first)",
      "",
      `Respond in your role as "${agent.name}". Keep it focused; the next agent will continue the thread.`,
    ].join("\n");
    const { text, errored } = await runOneTurn({
      spawn: ctx.spawnAgentTurn,
      agent,
      prompt,
      runId: ctx.runId,
      turn,
      phase: "none",
      bus: ctx.bus,
      totals,
      cwd: ctx.cwd,
      signal: ctx.signal,
      claudeBin: ctx.claudeBin,
    });
    if (errored) {
      return { status: "error", totals };
    }
    history.push({ agent: agent.name, text });
    const next = roster[turn % roster.length];
    if (next && next.name !== agent.name) {
      emitMessage(ctx.bus, agent.name, next.name, text);
    }
  }
  return { status: ctx.signal.aborted ? "aborted" : "completed", totals };
};

File: packages/claude-code-toolbox-intellij/src/main/kotlin/com/amitchorasiya/claude/toolbox/intellij/agents/runtime/protocols/RoundRobin.kt
package com.amitchorasiya.claude.toolbox.intellij.agents.runtime.protocols

import com.amitchorasiya.claude.toolbox.intellij.agents.runtime.*

fun roundRobin(ctx: ProtocolContext): ProtocolResult {
    val totals = RunUsage()
    val maxTurns = ctx.team.maxTurns
    val roster = ctx.agents
    if (roster.isEmpty()) return ProtocolResult("error", totals)

    val history = mutableListOf<Pair<String, String>>()

    for (turn in 1..maxTurns) {
        if (ctx.abortFlag.get()) return ProtocolResult("aborted", totals)
        val agent = roster[(turn - 1) % roster.size]
        val contextWindow = history.takeLast(6)
        val prompt = buildString {
            append("Team task: ${ctx.userPrompt}\n\n")
            if (contextWindow.isNotEmpty()) {
                append("Recent conversation:\n")
                for ((name, text) in contextWindow) append("[$name]: ${text.take(600)}\n\n")
            }
            append("You are \"${agent.name}\". Respond with your contribution.")
        }
        val result = spawnAgentTurn(agent, prompt, ctx.runId, turn, "none", ctx.bus, totals, ctx.cwd, ctx.claudeBin, ctx.abortFlag)
        history.add(agent.name to result.text)
        if (result.aborted) return ProtocolResult("aborted", totals)
    }
    return ProtocolResult("completed", totals)
}
